File: Scala/main/scala/com/hibiup/examples/Example_Call2Python.scala
package com.hibiup.examples

/**
  * 案例参考：https://www.py4j.org/advanced_topics.html#using-the-traditional-javagateway-gatewayserver
  */

/** *
  * 1) 定义一个用于 Python 的 interface
  *
  * 因为 Scala 不是动态语言，因此它需要预先知道 Python 方法的签名。这个签名将由 Python 来实现。参考 Python 脚本
  */
trait PythonHello {
    def sayHello(msg: String):String
}

import scala.concurrent.Future
import cats.effect.IO
import com.typesafe.scalalogging.Logger

/**
  * ClientServer (推荐)
  * */
object Example_Call2Python_With_ClientServer extends App{
    import py4j.ClientServer

    val logger = Logger(this.getClass)

    /**
      * 3) ClientServer 尝试连接 Python 端启动的 ClientServer 以获得注册的接口
      *
      * 　也可以用 GatewayServer，但是 Python 端必须改用 JavaGateway. 服务端 ClientServer 不支持 GatewayServer 交互。
      * */
    val gateway:ClientServer = new ClientServer()
    Future.successful(gateway.startServer())

    val service = IO{
        gateway.getPythonServerEntryPoint(Array[Class[_]](classOf[PythonHello])).asInstanceOf[PythonHello]
    }

    (for {
        io <- service
        res <- IO(io.sayHello("Java"))
    } yield res).unsafeRunAsync{
        case Right(msg) =>
            println(s"Receive message: $msg")
        case Left(t) =>
            logger.error(t.getMessage, t)
    }

    gateway.shutdown()
}

/**
  * GatewayServer
  * */
object Example_Call2Python_With_GatewayServer extends App{
    import py4j.GatewayServer

    /**
      * 3) GatewayServer 可以连接 Python 段启动的 JavaGateway 或 ClientServer 以获得注册的接口. 但是多次双向调用可能产生异常。
      * */
    val gateway:GatewayServer = new GatewayServer()
    Future.successful(gateway.start())  // 参数 fork 表示 gateway 是否运行在一个新线程中。

    // Very unsafe!! doing it with IO
    val service = gateway.getPythonServerEntryPoint(Array[Class[_]](classOf[PythonHello])).asInstanceOf[PythonHello]
    val msg = service.sayHello("Java")
    println(s"Receive message: $msg")

    gateway.shutdown()
}

File: PyServer/main/python/Example_Call2Python.py
"""
需要安装 PY4J:
~~~
$ python -m pip install py4j
~~~
"""

"""
1）实现  Java/Scala 的 interface/trait
"""


class PythonImpl(object):
    class Java:
        """
        声明实现的 interface
        """
        implements = ["com.hibiup.examples.PythonHello"]

    def sayHello(self, param):
        # 本地打印
        print("Python receive message: {0}".format(param))

        # GatewayServer 和 JavaGateway 具有双向交流能力，因此可以执行在客户端的输出。
        gateway.jvm.System.out.println("[" + gateway.jvm.Thread.currentThread().getName() + "] - " + param)  # JVM 打印

        # 返回给客户端
        return "Hello, {0}".format(param)


"""
2) 必须先于 JVM 启动 Python 程序.
"""
if __name__ == "__main__":
    # 生成 Python 端的服务实例
    impl = PythonImpl()

    # JavaGateway 会打开一个侦听端口用于接受来自客户端（ClientServer 或 GatewayServer）的注册请求。缺省端口是 25334
    # JavaGateway 似乎存在 Bug, 只能向 JVM 提交一次请求，因此建议使用 ClientServer
    from py4j.java_gateway import JavaGateway, CallbackServerParameters
    #ateway = JavaGateway(
    #    callback_server_parameters=CallbackServerParameters(),
    #    python_server_entry_point=impl  # 指定服务实例
    #)

    # ClientServer 可以实现多次双向交流，但是 ClientServer 的客户端也必须是 ClientServer，不能接入 GatewayServer
    from py4j.clientserver import ClientServer, JavaParameters, PythonParameters
    gateway = ClientServer(
        java_parameters=JavaParameters(),
        python_parameters=PythonParameters(),
        python_server_entry_point=impl)
